tests/test_database_types.py

```diff
@@ -0,0 +1,222 @@
+            # "integer",
+            # "bigint",
+            # "smallint",
+            # "tinyint",
+            # "byteint"
+        ],
+        # https://docs.snowflake.com/en/sql-reference/data-types-datetime.html
+        "datetime_no_timezone": [
+            "timestamp(0)",
+            "timestamp(3)",
+            "timestamp(6)",
+            "timestamp(9)",
+        ],
+        # https://docs.snowflake.com/en/sql-reference/data-types-numeric.html#decimal-numeric
+        "float": [
+            # "float"
+            # "numeric",
+        ],
+    },
+    db.Redshift: {
+        "int": [
+            # "int",
+        ],
+        "datetime_no_timezone": [
+            # "TIMESTAMP",
+        ],
+        # https://docs.aws.amazon.com/redshift/latest/dg/r_Numeric_types201.html#r_Numeric_types201-floating-point-types
+        "float": [
+            # "float4",
+            # "float8",
+            # "numeric",
+        ],
+    },
+    db.Oracle: {
+        "int": [
+            # "int",
+        ],
+        "datetime_no_timezone": [
+            # "timestamp",
+            # "timestamp(6)",
+            # "timestamp(9)",
+        ],
+        "float": [
+            # "float",
+            # "numeric",
+        ],
+    }
+}
+
+
+type_pairs = []
+# =>
+# { source: (preql, connection)
+# target: (preql, connection)
+# source_type: (int, tinyint),
+# target_type: (int, bigint) }
+for source_db, source_type_categories in DATABASE_TYPES.items():
+    for target_db, target_type_categories in DATABASE_TYPES.items():
+        for type_category, source_types in source_type_categories.items(): # int, datetime, ..
+            for source_type in source_types:
+                for target_type in target_type_categories[type_category]:
+                    if CONNS.get(source_db, False) and CONNS.get(target_db, False):
+                        type_pairs.append((
+                          source_db,
+                          target_db,
+                          source_type,
+                          target_type,
+                          type_category,
+                        ))
+
+# Pass --verbose to test run to get a nice output.
+def expand_params(testcase_func, param_num, param):
+    source_db, target_db, source_type, target_type, type_category = param.args
+    source_db_type = source_db.__name__
+    target_db_type = target_db.__name__
+    return "%s_%s_%s_to_%s_%s" %(
+        testcase_func.__name__,
+        source_db_type,
+        parameterized.to_safe_name(source_type),
+        target_db_type,
+        parameterized.to_safe_name(target_type),
+    )
+
+def _insert_to_table(conn, table, values):
+    insertion_query = f"INSERT INTO {table} (id, col) VALUES "
+    for j, sample in values:
+        insertion_query += f"({j}, '{sample}'),"
+
+    conn.query(insertion_query[0:-1], None)
+    if not isinstance(conn, db.BigQuery):
+        conn.query("COMMIT", None)
+
+class TestDiffCrossDatabaseTables(unittest.TestCase):
```

Comment:
I wonder if we should skip this one by default because it's so time-consuming?

Reply:
I still get some failures, for example, here's `snowflake timestamp(9) -> postgres timestamp(6)`, you can run with `-k test_types_Snowflake_timestamp_9__to_Postgres_timestamp_6_without_time_zone`.

```
======================================================================
FAIL: test_types_Snowflake_timestamp_9__to_Postgres_timestamp_6_without_time_zone (tests.test_database_types.TestDiffCrossDatabaseTables)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/simon/Library/Caches/pypoetry/virtualenvs/data-diff-DzlzoNC0-py3.9/lib/python3.9/site-packages/parameterized/parameterized.py", line 533, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/Users/simon/src/github.com/datafold/data-diff/tests/test_database_types.py", line 236, in test_types
    self.assertEqual(expected, diff)
AssertionError: Lists differ: [] != [('-', ('6', '2022-01-01 15:10:03.003031')[305 chars]0'))]

Second list contains 8 additional elements.
First extra element 0:
('-', ('6', '2022-01-01 15:10:03.003031'))

- []
+ [('-', ('6', '2022-01-01 15:10:03.003031')),
+  ('+', ('6', '2022-01-01 15:10:03.003030')),
+  ('-', ('7', '2022-01-01 15:10:05.009901')),
+  ('+', ('7', '2022-01-01 15:10:05.009900')),
+  ('-', ('8', '2022-01-01 15:10:05.009901')),
+  ('+', ('8', '2022-01-01 15:10:05.009900')),
+  ('-', ('9', '2022-01-01 15:10:05.009901')),
+  ('+', ('9', '2022-01-01 15:10:05.009900'))]
```

Seems to be:

```
test_types_Snowflake_timestamp_9__to_MySQL_timestamp_6_
test_types_Snowflake_timestamp_9__to_Postgres_timestamp_6_with_time_zone
test_types_Snowflake_timestamp_9__to_Postgres_timestamp_6_without_time_zone
```

Reply:
I don't know why, I don't get these failures.

Reply:
Yeah, it seems to only happen when Snowflake is the origin. Very, very weird it doesn't happen for you? Wtf?

Reply:
`poetry run python3 -m unittest tests/test_database_types.py  -v -k test_types_Snowflake_timestamp_9__to_Postgres_timestamp_6_without_time_zone`

Reply:
Sounds like lowercase/upper case shenanigans